doc/man/man8/strace.8.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man8/strace.8.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH STRACE 8 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Administration"
.\"
.\"
.SH NAME
.B strace
\- write \fISTREAMS\fR event trace messages to the standard output
.\"
.\"
.SH SYNOPSIS
.PP
.ad l
.nh
.HP \w'strace\ 'u
\fBstrace\fR [\fIoptions\fR]
[\fB-nqDv\fR]
[{\fB-d\fR|\fB--directory\fR[=]}\fIdirectory\fR]
[{\fB-p\fR|\fB--pidfile\fR[=]}[\fIpidfile\fR]]
[{\fB-l\fR|\fB--logdev\fR[=]}[\fIlogdevice\fR]]
[{\fB-D\fR|\fB--debug\fR[=]}[\fILEVEL\fR]]
[{\fB-v\fR|\fB--verbose\fR[=]}[\fILEVEL\fR]]
[\fImid\fR \fIsid\fR \fIlevel\fR] ...
.PD 0
.HP
\fBstrace\fR {\fB-h\fR|\fB--help\fR}
.HP
\fBstrace\fR {\fB-V\fR|\fB--version\fR}
.HP
\fBstrace\fR {\fB-C\fR|\fB--copying\fR}
.PD
.ad b
.hy 6
.\"
.\"
.SH DESCRIPTION
.PP
The
.B strace
receives trace event messages from the
.I STREAMS
log driver
.RB ( log (4))
and writes these messages to standard output.
When run as a daemon,
.B strace
appends these messages to a log file.
The error log files produced
reside in the directory
.BR /var/adm/streams " or " /var/log/streams ,
or a user-specified directory,
and are named
.BI trace. mm-dd\fR,
where
.I mm
is the month and
.I dd
is the day of the messages contained in each log file.
.PP
By default,
.B strace
without arguments writes all
.I STREAMS
trace messages from all drivers and modules.
.B strace
with command-line arguments limits the trace messages received.
.PP
Messages that appear in the trace log are intended to report debugging
information that assists with troubleshooting a running
.I STREAMS
module or driver.
Messages indicating the total failure of a
.I STREAMS
driver or module should have the
.B F
flag set.  Those messages requiring the immediate attention of the
administrator should have the
.B N
flag set, which causes the console logger to send the message to that person by
way of the
.BR mail (1)
command.  The priority level usually has no meaning in the error log, but does
have meaning if the message is sent to the trace process.  Messages of lower
priority levels can be filtered out.
.PP
Once initiated, the
.B strace
daemon continues to execute until terminated by the user.  Usually, the
.B strace
daemon is executed asynchronously.
.\"
.\"
.SH OPTIONS
.PP
.TP
\fB-d\fR, \fB--directory\fR \fIdirectory\fR
Specifies a directory to contain the messages.  Default is to search for
.RB \(lq /var/adm/streams \(rq
and then
.RB \(lq /var/log/streams .\(rq
.TP
\fB-n\fR, \fB--noforeground\fR
Daemonize, do not run in the foreground, print output to
.I outfile
and error logs to
.IR errfile .
This option is most usefull for debugging
.BR strace .
.TP
\fB-b\fR, \fB--basename\fR \fIbasename\fR
Specify the log file basename.  Default is
.RB \(lq trace \(rq .
.TP
\fB-o\fR, \fB--outfile\fR \fIoutfile\fR
Specifiy the log file output name.
When run as a daemon, the default is
.RI \(lq basename . mm-dd \(rq,
where
.I mm-dd
is the current month and day of month.
When run in the foreground, the default is ignored and all output is to
.IR stdout .
When specified, standard output is redirected to the specified file regardless of whether
.B strace
is being run in the foreground or as a daemon.
.TP
\fB-e\fR, \fB--errfile\fR \fIerrfile\fR
Specify the error file output name.
When run as a daemon, the default is
.RI \(lq /dev/null \(rq
and all errors are logged to
.BR syslogd (8).
When run in the foreground, the default is ignored and all errors are logged to
.IR stderr .
When specified, standard error is redirected to the specified file regardless of whether
.B strace
is being run in the foreground or as a daemon.
.TP
\fB-p\fR, \fB--pidfile\fR \fIpidfile\fR
Specify the pid file output name.  Default is
.RB \(lq /var/run/strace.pid .\(rq
When run in the foreground, the default is ignored.
.TP
\fB-l\fR, \fB--logdev\fR \fIlogdevice\fR
Specifies the log device.  Default is to search
.RB \(lq /dev/streams/strlog ,\(rq
.RB \(lq /dev/streams/log ,\(rq
.RB \(lq /dev/strlog \(rq
and
.RB \(lq /dev/log .\(rq
.TP
\fB-q\fR, \fB--quiet\fR
Suppresses normal output.
This option is equivalent to
.BR --verbose = 0 .
.TP
\fB-D\fR, \fB--debug\fR [\fIlevel\fR]
Increase or specify the debug output
.IR level .
When a debug
.I level
has been specified,
.B strace
will log to
.IR stdout ,
prints debug messages to
.IR stderr ,
and will not place itself in the background.
This option may be repeated.
.TP
\fB-v\fR, \fB--verbose\fR [\fIlevel\fR]
Increase or specify the verbosity
.I level
of the output.  This option may be repeated.
.TP
\fB-h\fR, \fB--help\fR
When this option is encountered,
.B strace
displays usage information on
.I stdout
and exits.
No further processing of options or arguments is performed and no actions are taken.
.TP
\fB-V\fR, \fB--version\fR
When this option is encountered,
.B strace
displays version information on
.I stdout
and exits.
No further processing of options or arguments is performed and no actions are taken.
.TP
\fB-C\fR, \fB--copying\fR
When this option is encountered,
.B strace
displays copying and copyright information on
.I stdout
and exits.
No further processing of options or arguments is performed and no actions are taken.
.\"
.\"
.SH ARGUMENTS
.PP
The arguments, which must be specified in groups of three, are:
.TP
.I mid
Specifies the
.I STREAMS
module identification number from the
.B streamtab
entry.
.TP
.I sid
Specifies a sub-identification number (often corresponding to a minor device).
.TP
.I level
Specifies a tracing priority level.
.B strace
gets messages of a level equal to or less than the value specified by
.BR level .
Only positive integer values are allowed.
.PP
The value
.RI \*(lq all \*(rq
can be used for any argument in the
.B strace
command line to indicate that there are no restrictions for that argument.
.PP
Multiple sets of the three arguments can be specified to obtain the messages
from more than one driver or module.
When
.B strace
is invoked, the log driver compares the sets of command line arguments with
actual trace message, returning only messages that satisfy the specified
criteria.
.\"
.\"
.SH FORMAT
.PP
.I STREAMS
trace event messages have the following format:
.HP
.RB < seq "> <" time "> <" ticks "> <" level "> <" flags "> <" mid "> <" sid "> <" text >
.PP
Components are interpreted as follows:
.TP \w'<flags>\(em'u
.RB < seq >
Trace sequence number.
.PD 0
.TP
.RB < time >
Time the message was sent expressed in hh:mm:ss.
.TP
.RB < ticks >
Time the message was sent expressed in machine ticks since last boot.
.TP
.RB < level >
Tracing level as defined by the
.I STREAMS
driver or module that originates the messages.
.TP
.RB < flags >
Any combination of the following three message indicators:
.RS
.IP \fBE\fP 3
The message has also been saved in the error log.
.IP \fBF\fP

NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.


src/util/init/strace.sh
#!/bin/sh
#
# @(#) src/util/init/strace.sh
# Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
# Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
# Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
# All Rights Reserved.
#
# Distributed by OpenSS7 Corporation.  See the bottom of this script for copying
# permissions.
#
# These are arguments to update-rc.d ala chkconfig and lsb.  They are recognized
# by openss7 install_initd and remove_initd scripts.  Each line specifies
# arguments to add and remove links after the the name argument:
#
# strace:	start and stop strace facility
# update-rc.d:	start 35 S . stop 35 0 6 .
# lockfile:	/var/lock/subsys/strace
# config:	/etc/default/strace
# processname:	strace
# pidfile:	/var/run/strace.pid
# probe:	true
# hide:		false
# license:	AGPL
# description:	This STREAMS init script is part of Linux Fast-STREAMS.  It is \
#		responsible for starting and stopping the STREAMS trace \
#		logger.  The STREAMS trace logger should only be run under \
#		exceptional circumstances and this init script not activated \
#		automatically.
#
# LSB init script conventions
#
### BEGIN INIT INFO
# Provides: strace
# Required-Start: streams
# Required-Stop: streams
# Should-Start: $syslog strerr
# Should-Stop: $syslog strerr
# Default-Start: S
# Default-Stop:
# X-UnitedLinux-Default-Enabled: no
# Short-Description: start and stop strace
# License: AGPL
# Description:	This STREAMS init script is part of Linux Fast-STREAMS.  It is
#	responsible for starting and stopping the STREAMS trace logger.  The
#	STREAMS trace logger should only be run under exceptional circumstances
#	and this init script is not activated in any run level by default.
### END INIT INFO

PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin

# Source init script functions library.
init_mode="standalone"
  if [ -r /etc/init.d/functions   ] ; then . /etc/init.d/functions   ; init_mode="chkconfig" ;
elif [ -r /etc/rc.status          ] ; then . /etc/rc.status          ; init_mode="insserv"   ;
elif [ -r /lib/lsb/init-functions ] ; then . /lib/lsb/init-functions ; init_mode="lsb"       ;
elif [ -r /sbin/start-stop-daemon ] ; then                             init_mode="debian"    ;
fi

case "$init_mode" in
    (chkconfig)
	# RedHat's LSB support is not so good, use native functions first.
	daemon_start () {
	    local force= nicelevel= base=
	    while [ "$1" != "${1##[-+]}" ] ; do
		case $1 in
		    (-f) force="--force" ; shift 1 ;;
		    (-n) nicelevel="$2"  ; shift 2 ;;
		esac
	    done
	    base=`echo $1 | sed -e 's,.*/,,'`
	    echo -en "Starting daemon $base: "
	    daemon $force $nicelevel "$@"
	    RETVAL=$?
	    echo ""
	    return $RETVAL
	}
	daemon_stop () {
	    local base=
	    base=`echo $1 | sed -e 's,.*/,,'`; shift 1
	    echo -en "Stopping daemon $base: "
	    killproc $base "$@"
	    RETVAL=$?
	    echo ""
	    return $RETVAL
	}
	daemon_reload () {
	    local base=
	    base=`echo $1 | sed -e 's,.*/,,'`; shift 1
	    echo -en "Reloading daemon $base: "
	    killproc $base -HUP
	    RETVAL=$?
	    echo ""
	    return $RETVAL
	}
	;;
    (insserv)
	# SuSE's LSB support is not so good, use native functions first.
	daemon_start () {
	    local force= nicelevel= base=
	    while [ "$1" != "${1##[-+]}" ] ; do
		case $1 in
		    (-f) force="-f" ; shift 1 ;;
		    (-n) nicelevel="-n $2"  ; shift 2 ;;
		esac
	    done
	    base=`echo $1 | sed -e 's,.*/,,'`
	    echo -en "Starting daemon $base: "
	    startproc $force $nicelevel "$@"
	    RETVAL=$?
	    [ $RETVAL -eq 0 ] || rc_failed $RETVAL
	    rc_status -v
	    return $RETVAL
	}
	daemon_stop () {
	    local base=
	    base=`echo $1 | sed -e 's,.*/,,'`; shift 1
	    echo -en "Stopping daemon $base: "
	    killproc $base "$@"
	    RETVAL=$?
	    [ $RETVAL -eq 0 ] || rc_failed $RETVAL
	    rc_status -v
	    return $RETVAL
	}
	daemon_reload () {
	    local base=
	    base=`echo $1 | sed -e 's,.*/,,'`; shift 1
	    echo -en "Reloading daemon $base: "
	    killproc $base -HUP
	    RETVAL=$?
	    [ $RETVAL -eq 0 ] || rc_failed $RETVAL
	    rc_status -v
	    return $RETVAL
	}
	action () {
	    echo -en "$1"
	    shift 1
	    ${1+"$@"} >/dev/null
	    RETVAL=$?
	    [ $RETVAL -eq 0 ] || rc_failed $RETVAL
	    rc_status -v
	    return $RETVAL
	}
	;;
    (lsb)
	daemon_start () {
	    local force= nicelevel= base=
	    while [ "$1" != "${1##[-+]}" ] ; do
		case $1 in
		    (-f) force="-f" ; shift 1 ;;
		    (-n) nicelevel="-n $2"  ; shift 2 ;;
		esac
	    done
	    base=`echo $1 | sed -e 's,.*/,,'`
	    echo -en "Starting daemon $base: "
	    start_daemon $force $nicelevel "$@"
	    RETVAL=$?
	    [ $RETVAL -eq 0 ] && log_success_msg || log_failure_msg
	    return $RETVAL
	}
	daemon_stop () {
	    local base=
	    base=`echo $1 | sed -e 's,.*/,,'`; shift 1
	    echo -en "Stopping daemon $base: "
	    killproc $base "$@"
	    RETVAL=$?
	    [ $RETVAL -eq 0 ] && log_success_msg || log_failure_msg
	    return $RETVAL
	}
	daemon_reload () {
	    local base=
	    base=`echo $1 | sed -e 's,.*/,,'`; shift 1
	    echo -en "Reloading daemon $base: "
	    killproc $base -HUP
	    RETVAL=$?
	    [ $RETVAL -eq 0 ] && log_success_msg || log_failure_msg
	    return $RETVAL
	}
	action () {
	    echo -en "$1"
	    shift 1
	    ${1+"$@"} >/dev/null
	    RETVAL=$?
	    [ $RETVAL -eq 0 ] && log_success_msg || log_failure_msg
	    return $RETVAL
	}
	;;
    (debian)
	# a resonably modern debian system will have LSB support
	daemon_start () {
	    local force= nicelevel= base=
	    while [ "$1" != "${1##[+-]}" ] ; do
		case $1 in
		    (-f) force="" ; shift 1 ;;
		    (-n) nicelevel="-N $2"  ; shift 2 ;;
		esac
	    done
	    base=`echo $1 | sed -e 's,.*/,,'`
	    echo -en "Starting daemon $base: "
	    start-stop-daemon --start $force $nicelevel --quiet \
		--pidfile $pidfile --exec $execfile -- "$@"
	    RETVAL=$?
	    [ $RETVAL -eq 0 ] || echo "(failed)"; echo "."
	    return $RETVAL
	}
	daemon_stop () {
	    local base=
	    base=`echo $1 | sed -e 's,.*/,,'`; shift 1
	    echo -en "Stopping daemon $base: "
	    start-stop-daemon --stop --quiet --retry=1 --oknodo \
		--pidfile $pidfile --exec $execfile
	    RETVAL=$?
	      if [ $RETVAL -eq   0 ] ; then echo "."
	    elif [ $RETVAL -eq 255 ] ; then echo "(warning)."
	    else                            echo "failed!"
	    fi
	    return $RETVAL
	}
	daemon_reload () {
	    local base=
	    base=`echo $1 | sed -e 's,.*/,,'`; shift 1
	    echo -en "Reloading daemon $base: "
	    killproc $base -HUP
	    RETVAL=$?
	    [ $RETVAL -eq 0 ] || echo "(failed)"; echo "."
	    return $RETVAL
	}
	action () {
	    echo -en "$1"
	    shift 1
	    eval "\${1+\"\$@\"} $redir"
	    RETVAL=$?
	    [ $RETVAL -eq 0 ] || echo "(failed)"; echo "."
	    return $RETVAL
	}
	;;
    (standalone|*)
	if [ -x /sbin/pidof ] ; then
	    pidofproc () {
		/sbin/pidof $@
	    }
	elif [ -x /sbin/pidofproc ] ; then
	    pidofproc () {
		/sbin/pidofproc $@
	    }
	else
	    pidofproc () {
		[ -r $pidfile ] && cat $pidfile
	    }
	fi
	if [ -x /sbin/start-stop-daemon ] ; then
	    killproc () {
		local base execfile
		base=$1
		execfile=$2
		shift 2
		start-stop-daemon --stop --quiet --retry=1 --oknodo \
		    --pidfile $pidfile \
		    --exec $execfile
	    }
	elif [ -x /sbin/killproc ] ; then
	    killproc () {
		/sbin/killproc $@
	    }
	elif [ -x /usr/bin/killall ] ; then
	    killproc () {
		/usr/bin/killall $@
	    }
	else
	    killproc () {
		[ -r $pidfile ] && kill "$2" `cat $pidfile`
	    }
	fi
	daemon_start () {
	    local force= nicecmd= base=
	    while [ "$1" != "${1##[-+]}" ] ; do
		case $1 in
		    (-f) force="--force"       ; shift 1 ;;
		    (-n) nicecmd="nice -n $2"  ; shift 2 ;;
		esac
	    done
	    base=`echo $1 | sed -e 's,.*/,,'`
	    echo -en "Starting daemon $base: "
	    eval "( $nicecmd \"\$@\" >/dev/null 2>&1 </dev/null & )"
	    RETVAL=$?
	    [ $RETVAL -eq 0 ] && echo -en "\t...SUCCESS" || echo -en "\t....FAILED"
	    return $RETVAL
	}
	daemon_stop () {
	    local base=
	    base=`echo $1 | sed -e 's,.*/,,'`; shift 1
	    echo -en "Stopping daemon $base: "
	    if pidofproc $base >/dev/null 2>&1 ; then
		RETVAL=0
	    else
		killproc $base "$@"
		RETVAL=$?
	    fi
	    [ $RETVAL -eq 0 ] && log_success_msg || log_failure_msg
	    return $RETVAL
	}
	daemon_reload () {
	    local base=
	    base=`echo $1 | sed -e 's,.*/,,'`; shift 1
	    echo -en "Reloading daemon $base: "
	    killproc $base -HUP
	    RETVAL=$?
	    [ $RETVAL -eq 0 ] && echo -en "\t...SUCCESS" || echo -en "\t....FAILED"
	    return $RETVAL
	}
	action () {
	    echo -en "$1"
	    shift 1
	    ${1+"$@"} >/dev/null
	    RETVAL=$?
	    [ $RETVAL -eq 0 ] && echo -e "\t...SUCCESS" || echo -e "\t....FAILED"
	    return $?
	}
	;;
esac

name='strace'
script='strace.sh'
ucname="STRACE"
lockfile="/var/lock/subsys/$name"
config="/etc/sysconfig/$name /etc/default/$name"
processname="$name"
pidfile="/var/run/$processname.pid"
execfile="/usr/sbin/$processname"
desc="the STREAMS trace logger"

[ -x $execfile -o "$1" = "stop" ] || exit 5

# Specify defaults

STRACEOPTIONS=
STRACE_DIRECTORY="/var/log/streams"
STRACE_BASENAME="trace"
STRACE_OUTFILE=
STRACE_ERRFILE=
STRACE_LOGDEVICE="/dev/streams/clone/log"
STRACE_MODULES=
STRACE_OPTIONS=

# Source redhat and/or debian config file
for file in $config ; do
    [ -f $file ] && . $file
done

RETVAL=0

umask 077

case ":$VERBOSE" in
    :no|:NO|:false|:FALSE|:0|:)
	redir='>/dev/null 2>&1'
	;;
    *)
	redir=
	;;
esac

build_options () {
    # Build up the options string
    STRACE_OPTIONS="-n -p $pidfile"
    [ -n "$STRACEOPTIONS" ] && \
	STRACE_OPTIONS="${STRACE_OPTIONS}${STRACE_OPTIONS:+ }${STRACEOPTIONS}"
    [ -n "$STRACE_DIRECTORY" ] && \
	STRACE_OPTIONS="${STRACE_OPTIONS}${STRACE_OPTIONS:+ }-d ${STRACE_DIRECTORY}"
    [ -n "$STRACE_BASENAME" ] && \
	STRACE_OPTIONS="${STRACE_OPTIONS}${STRACE_OPTIONS:+ }-b ${STRACE_BASENAME}"
    [ -n "$STRACE_OUTFILE" ] && \
	STRACE_OPTIONS="${STRACE_OPTIONS}${STRACE_OPTIONS:+ }-o ${STRACE_OUTFILE}"
    [ -n "$STRACE_ERRFILE" ] && \
	STRACE_OPTIONS="${STRACE_OPTIONS}${STRACE_OPTIONS:+ }-e ${STRACE_ERRFILE}"
    [ -n "$STRACE_LOGDEVICE" ] && \
	STRACE_OPTIONS="${STRACE_OPTIONS}${STRACE_OPTIONS:+ }-l ${STRACE_LOGDEVICE}"
    [ -n "$STRACE_MODULES" ] && \
	STRACE_OPTIONS="${STRACE_OPTIONS}${STRACE_OPTIONS:+ }${STRACE_MODULES}"
}

start () {
    [ -d "$STRACE_DIRECTORY" -o -z "$STRACE_DIRECTORY" ] || \
	action $"Creating $STRACE_DIRECTORY: " \
	    mkdir -p -- "$STRACE_DIRECTORY" \
	    || RETVAL=$?
    build_options
    daemon_start $execfile $STRACE_OPTIONS \
	|| RETVAL=$?
    [ $RETVAL -eq 0 ] && touch $lockfile
    return $RETVAL
}

stop () {
    if [ -r $pidfile ]; then
	daemon_stop $execfile || RETVAL=$?
    else
	RETVAL=0
    fi
    [ $RETVAL -eq 0 ] && rm -f -- $lockfile
    return $RETVAL
}

restart () {
    stop  || RETVAL=$?
    start || RETVAL=$?
    return $RETVAL
}

reload () {
    daemon_reload $execfile || RETVAL=$?
    return $RETVAL
}

show () {
    echo "$script: show: not yet implemented." >&2
    RETVAL=1
    return $RETVAL
}

usage () {
    echo "Usage: $script (start|stop|status|restart|try-restart|condrestart|force-reload|reload|probe|show)" >&2
    RETVAL=1
    return $RETVAL
}

case "$1" in
    (start|stop|reload|restart|show)
	$1 || RETVAL=$?
	;;
    (force-reload)
	reload || RETVAL=$?
	;;
    (status)
	$1 $name || RETVAL=$?
	;;
    (try-restart|condrestart)
	[ -f $lockfile ] && restart || RETVAL=$?
	;;
    (probe)
	if pidofproc $processname >/dev/null 2>&1 ; then
	    # running
	    if [ "`pidofproc $processname`" -eq "`cat $pidfile`" ] ; then
		# running with the right pid
		if [ ! -f $lockfile ] ; then
		    # running but subsystem unlocked, need to reload
		    echo 'reload'
		else
		    # subsystem locked
		    for file in $config ; do
			if [ -f $file -a $file -nt $lockfile ] ; then
			    # configuration file updated, need to reload
			    echo 'reload'
			    break
			fi
		    done
		fi
	    else
		# running, but with the wrong pid, need to restart
		echo 'restart'
	    fi
	else
	    # not running
	    if [ ! -f $lockfile ] ; then
		# subsystem unlocked, need to start
		echo 'start'
	    else
		# dead but subsystem locked, need to restart
		echo 'restart'
	    fi
	fi
	# do not need to do anything
	RETVAL=$?
	;;
    (*)
	usage || RETVAL=$?
	;;
esac

[ "${0##*/}" = "$script" ] && exit $RETVAL

# =============================================================================
# 
# @(#) src/util/init/strace.sh
#
# -----------------------------------------------------------------------------
#
# Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
# Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
# Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
#
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation; version 3 of the License.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program.  If not, see <http://www.gnu.org/licenses/>, or write to
# the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#
# -----------------------------------------------------------------------------
#
# U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
# behalf of the U.S. Government ("Government"), the following provisions apply
# to you.  If the Software is supplied by the Department of Defense ("DoD"), it
# is classified as "Commercial Computer Software" under paragraph 252.227-7014
# of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
# successor regulations) and the Government is acquiring only the license rights
# granted herein (the license rights customarily provided to non-Government
# users).  If the Software is supplied to any unit or agency of the Government
# other than DoD, it is classified as "Restricted Computer Software" and the
# Government's rights in the Software are defined in paragraph 52.227-19 of the
# Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
# the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
# (or any successor regulations).
#
# -----------------------------------------------------------------------------
#
# Commercial licensing and support of this software is available from OpenSS7
# Corporation at a fee.  See http://www.openss7.com/
#
# =============================================================================

# vim: ft=sh sw=4 tw=80
